File: client/src/api/systemPrompt.js
/**
 * api/systemPrompt.js
 * -------------------
 * The live system prompt + its version history - see
 * server/routes/systemPrompt.js.
 */

import { apiFetch, cachedGet, invalidateCache } from "./client.js";

/** Rarely changes; 5-min TTL so switching to this tab and back is instant. saveSystemPrompt() invalidates it. */
export async function fetchSystemPrompt() {
  return cachedGet("/api/system-prompt", 5 * 60 * 1000);
}

export async function saveSystemPrompt(content) {
  const result = await apiFetch("/api/system-prompt", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
  });
  invalidateCache("/api/system-prompt"); // both the prompt itself and its version list
  return result;
}

/** Every past system-prompt version, newest first, PREVIEW TEXT ONLY (content_preview/content_length, not the full content) - see fetchSystemPromptVersionContent for the full text of one version. */
export async function fetchSystemPromptVersions() {
  return cachedGet("/api/system-prompt/versions", 5 * 60 * 1000);
}

/** One version's full text - call only when a History card is expanded or right before Restore, never for the list. */
export async function fetchSystemPromptVersionContent(versionId) {
  return apiFetch(`/api/system-prompt/versions/${versionId}`);
}

File: pipeline/routers/system_prompt.py
"""routers/system_prompt.py - read/write the live system prompt + its version history."""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from auth_dependency import require_auth
from system_prompt_store import (
    load_system_prompt,
    save_system_prompt,
    list_system_prompt_versions,
    get_system_prompt_version_content,
)

router = APIRouter()


class SystemPromptPayload(BaseModel):
    content: str


@router.get("/system-prompt", dependencies=[Depends(require_auth)])
def get_system_prompt():
    """Return the system prompt exactly as it is used for the next generation."""
    return {"content": load_system_prompt()}


@router.put("/system-prompt", dependencies=[Depends(require_auth)])
def update_system_prompt(payload: SystemPromptPayload):
    """
    Save the edit as a new version in MongoDB. Because draft_generator.py
    reads the latest version fresh on every call, this takes effect on the
    very next /generate request - no restart needed. Every past version
    stays in the database.
    """
    version = save_system_prompt(payload.content)
    return {"status": "saved", "version": version}


@router.get("/system-prompt/versions", dependencies=[Depends(require_auth)])
def system_prompt_versions():
    """Every past version, newest first, preview text only - full content is a separate on-demand call (see below)."""
    return {"versions": list_system_prompt_versions()}


@router.get("/system-prompt/versions/{version_id}", dependencies=[Depends(require_auth)])
def system_prompt_version_content(version_id: str):
    """One version's full text - called only when a History card is expanded, or right before Restore."""
    result = get_system_prompt_version_content(version_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Version not found")
    return result
